### DISCUSSION TITLE
v2.9.378.0

### DISCUSSION THREAD

**Discussion by doug24** [dnGrep/dnGrep#549] [General]
## Changes in this release:

Two big enhancements and six smaller changes in this release:

#520 **Restore from maximized**:  added another change to set the preview panel size when restoring.

#528 **Simple Translate Option**:  added the capability to translate dnGrep into different languages by moving all the text into resource assemblies, and dynamically switch languages with a selection on the Options dialog.  This release only contains an English language resource, but more will be added when translations are completed.  This is where you can help: if you want to see dnGrep in your language, please help with a translation.  See [here for more information](https://github.com/dnGrep/dnGrep/discussions/538).  With this release you can create a translation by editing an XML file, and loading it into the running application to see the results.  

#529 **Check for valid source folder**:  moved the check for valid source paths to when the `Search` button is clicked.  Previously the `Search` button would be disabled if the source folder is invalid.

#530 **Add hot keys for Search and Replace fields**.  Alt+f and Alt+p will change the input focus to these fields.  Also added F6 and Shift+F6 to expand all and collapse all search results.

#539 **Word document disable macros**:  In the old style Word document (*.doc) search plugin, dnGrep disables running any macros when searching the documents.

#541 **Detect file encoding**:  fixed a bug when detecting the file encoding when search for files rather than searching within files.  If the file is not found, the error is now captured and reported the same way as when searching within files.

#542 **Search for Hex string**:  added a new feature to search files in binary mode.  The source files may be either text or binary.  The search string is one or more hexadecimal numbers representing byte values (00 to ff inclusive).   The hex numbers are space delimited:  `68 65 78 20`.  A wildcard, either `?` or `??`, will match one byte of any value.  The search string is converted to a byte array and the file is searched in 4kb blocks for matching sequences of bytes.  Search results will not overlap.  For example, if you search for 2 null values (`00 00`) a sequence of three consecutive null values in the file will match once on the first two nulls, and not a second time on the second two nulls. A sequence of four null values will match twice, not three times.  The search results are shown as hex values: formatted with 16 bytes per line.  The line number field on the left is the offset from the start of the file to the first byte on the line.  To the right of the 16 byte values is a text representation of the bytes on that line, if they are printable values.  To convert the bytes to text, the file encoding is used if dnGrep is in auto-encoding mode, or the user set encoding.  If the file is binary, the system default encoding is used if the user has not selected a specific encoding.  

#545 **Pass date fields to Everything search**:  if the Everything search service is installed and dnGrep is set to use Everything, then dnGrep will append the date and file size filters to the Everything search string so that filtering will be done by Everything and not dnGrep.  There was a bug in dnGrep that prevented using those filters with Everything searches.

### Checksum information

Name: dnGREP.2.9.378.x64.msi
Size: 6210683 bytes (6065 KiB)
SHA256: E1BF344BA90F4B0100839BCEE0F62E81D5463E246914613639A2DE24B6E8C0A8

Name: dnGREP.2.9.378.x86.msi
Size: 6202489 bytes (6057 KiB)
SHA256: E882AE81A16A91ECF76A0FCC209CADC210A79A02A6A3FF2503CAD2D3DD6512F6

<hr /><em>This discussion was created from the release <a href='https://github.com/dnGrep/dnGrep/releases/tag/v2.9.378.0'>v2.9.378.0</a>.</em>

**Comment by GflowerG** [comment]
**Hex strings works great!** Now dnGrep is the leader of such tools for me...

Only one request: show search results formatted 16/32/64/128 bytes per line, optional.

**Reply by doug24** [reply]
This looks like making the line length configurable should be possible.  I just ran a quick check changing it from 16 to 32.  I suspect that longer than 32 will be difficult to read, however.  I'll come back to this when the translation changes have slowed down.
